rest/route/host_agent.go

```diff
@@ -483,6 +485,71 @@ func assignNextAvailableTask(ctx context.Context, env evergreen.Environment, tas
 			return nil, true, nil
 		}
 
+		if currentHost.Distro.SingleTaskDistro {
+			// If next task exists and the distro is a single task distro, check if the task is allowed on the distro.
+			allowedTasks, err := validator.GetAllowedSingleTaskDistroTasksForProject(ctx, nextTask.Project)
+			if err != nil {
+				errMsg = message.Fields{
+					"message":    "could not find allowed single task disto tasks for project",
+					"host_id":    currentHost.Id,
+					"distro_id":  nextTask.DistroId,
+					"task_id":    nextTask.Id,
+					"task_name":  nextTask.DisplayName,
+					"task_group": nextTask.TaskGroup,
+					"project":    projectRef.Id,
+				}
+				grip.Alert(message.WrapError(err, errMsg))
+				return nil, false, errors.Wrapf(err, "could not find allowed single task disto tasks for project '%s'", nextTask.Project)
+			}
+			matched := false
+			for _, allowedTask := range allowedTasks {
+				matched, err = regexp.MatchString(allowedTask, nextTask.DisplayName)
+				if err != nil {
+					errMsg = message.Fields{
+						"message":      "could not process regex",
+						"task_id":      nextTask.Id,
+						"task_name":    nextTask.DisplayName,
+						"allowed_task": allowedTask,
+					}
+					grip.Alert(message.WrapError(err, errMsg))
+					return nil, false, errors.Wrapf(err, "could not process regex '%s'", allowedTask)
+				}
+				if matched {
+					break
+				}
+			}
+			if !matched {
+				grip.Alert(message.Fields{
+					"message":            "top task queue task is not allowed on single task distros",
+					"host_id":            currentHost.Id,
+					"distro_id":          nextTask.DistroId,
+					"task_id":            nextTask.Id,
+					"task_name":          nextTask.DisplayName,
+					"task_group":         nextTask.TaskGroup,
+					"project":            projectRef.Id,
+					"project_identifier": projectRef.Enabled,
+				})
+
+				// Mark task as system failed if it is not allowed on the distro.
+				details := &apimodels.TaskEndDetail{
+					Status:      evergreen.TaskSystemFailed,
+					Description: "Task not allowed to run on single task distros",
+				}
+
+				err = model.MarkEnd(ctx, env.Settings(), nextTask, evergreen.APIServerTaskActivator, time.Now(), details, false)
+				if err != nil {
+					errMsg = message.Fields{
+						"message":   "could not mark task as system failed",
```

Comment:
nit: might help to clarify in the log message that the reason we're marking it system failed is because it's not on the allowed task list.